### PR TITLE
Add initial benchmarking setup.

### PR DESCRIPTION
**Pull request by jerzywilczek**
Here's an initial benchmarking setup. For context, the end goal for this is for us to be able to run a set of these benchmarks on any computer and figure out what kind of smelter setups can be run on it. 

You can run it with:
```sh
cargo run --bin benchmark -- --help
```
to get the information on how to run a single test.

The full command is quite verbose, and looks like this:
```sh
cargo run --release --bin benchmark -- \
    --framerate 24 \
    --decoder-count maximize \
    --file-path examples/assets/BigBuckBunny.mp4 \
    --output-width 1280 \
    --output-height 720 \
    --encoder-preset ultrafast \
    --warm-up-time 10 \
    --measured-time 10 \
    --video-decoder vulkan_video_h264 \
    --framerate-tolerance 1.05
```
The help message will tell you all of the options for each of the arguments.

It can currently run a single test configuration. A single configuration is comprised of a couple of constant parameters (the resolution, the input video, the encoder preset etc.) and currently two variable parameters: framerate and decoders count. The variable parameters can be set to:
- a constant
- exponential iteration, meaning that the test will run a couple of iterations with the parameter being multiplied by 2 in every consecutive run, until the test stops running
- maximizing, meaning that the benchmark will binsearch the maximum number the configuration works for

Importantly, you can set multiple parameters to iterate and a single parameter to maximize, which would mean the setup would run for all iterative combinations and for each of them find the maximum number for the maximized parameter.

For example, if we set the framerate to iterate and the decoders count to maximize, the setup will run with framerate set to 1, 2, 4, 8, 16 etc. and for each framerate find the maximum number of inputs. The iteration will terminate when it is impossible to run the system due to the super high framerate even with a single decoder.

What I plan to add to this later:
- more parameters that can be variable: encoder count, encoder preset, resolution etc.
- a possibility to run without one of the elements of the pipeline, e.g. just the decoders and the renderer, without the encoder, or just the encoders, without the decoders
- a predefined set of benchmark runs that can be run with a single command, once we decide on what we want to measure

Additionally to asking whether you like this system or think it should be changed somehow, I would like to get your opinion on my handling of the arguments that are set to iterate or maximize (the `run_args` function and the other functions it calls). I feel a bit like using zig for the last couple weeks got into my head and that there is a better, more idiomatic way of doing what I'm doing here, but I just can't figure it out.

I also fixed an unrelated typo in vk-video, because it was annoying me.

### CODE REVIEW

File: integration_tests/src/bin/benchmark/main.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                })),
+            },
+        },
+    )
+    .unwrap();
+
+    Pipeline::start(&pipeline);
+
+    let start_time = Instant::now();
+    while Instant::now() - start_time < bench_config.warm_up_time {
```
Comment (wkozyra95) [inline integration_tests/src/bin/benchmark/main.rs:286]:
```suggestion
    while start_time.elapsed() < bench_config.warm_up_time {
```

Comment (wkozyra95) [reply integration_tests/src/bin/benchmark/main.rs:286]: alternative approach would be to just do thread sleep and reading from receiver in a separate channel

**Review by wkozyra95 — APPROVED** [review]

File: integration_tests/src/bin/benchmark/main.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+                    source: Source::File(bench_config.file_path.clone()),
+                }),
+                queue_options: QueueInputOptions {
+                    offset: Some(Duration::ZERO),
+                    required: true,
+                    buffer_duration: None,
+                },
+            },
+        )
+        .unwrap();
```
Comment (brzep) [inline integration_tests/src/bin/benchmark/main.rs:234]: I came across an error 
```
InputError(InputId("input_121"), Mp4(IoError(Os { code: 24, kind: Uncategorized, message: "Too many open files" })))
```
It should be handled somehow imo. Maybe just log it and assume that last count was max? I think it would be pretty anoying when I got such error after running test for longer time.

Maybe make `run_single_test` return `Result<bool, some-kind-of-error>` and handle those on higher level?

File: integration_tests/src/bin/benchmark/main.rs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                    continue;
+                } else {
+                    return any_succeeded;
+                }
+            }
+        }
+    }
+
+    run_args_maximize(ctx, args, arguments, reports)
+}
```
Comment (noituri) [inline integration_tests/src/bin/benchmark/main.rs:115]: I'm not sure about this code. On the first sight I thought there might be an infinite recursion. While now I understand what's happening here, the flow here is hard to follow. Also, I tried to run `--decoder-count` with `iterate_exp` and it didn't do anything. I'm not sure if I used it correctly tho

**Review by noituri — COMMENTED** [review]
🤓

File: integration_tests/src/bin/benchmark/main.rs (hunk 2 above)
Comment (jerzywilczek) [reply integration_tests/src/bin/benchmark/main.rs]: fixed it quickly for now

File: integration_tests/src/bin/benchmark/main.rs (hunk 3 above)
Comment (jerzywilczek) [reply integration_tests/src/bin/benchmark/main.rs:115]: you must have not set any argument to `maximize`, at which point the test is kind of pointless. This should be caught and reported as a problem: there's nothing you can learn by just setting all arguments to `iterate` and none to maximize. If you set framerate to `iterate_exp` and decoders to `maximize` it will work.

I'll add code to catch this sort of a mistake.

Comment (jerzywilczek) [reply integration_tests/src/bin/benchmark/main.rs]: I also added comments in some places, because I don't have ideas on how to make this easier to read. I also changed the return type of this function to `struct FurtherIterationPossible(bool)` to make what the returned `bool` means.

**Review by noituri — APPROVED** [review]
